### ISSUE TITLE
Blocking Call to Open inside Event Loop

### ISSUE THREAD

**Issue by nima-1102** [jaydeethree/Home-Assistant-weatherdotcom#33]
A blocking call to open was detected inside the event loop by the custom integration 'weatherdotcom'. This issue occurs in the file coordinator.py at line 232 within the custom component 'weatherdotcom'. The problem arises when attempting to load JSON data from a file, which results in the event loop being blocked.

**Steps to Reproduce:**

Install and configure the 'weatherdotcom' custom integration.
Start the Home Assistant service.
Observe the log for errors related to blocking calls within the event loop.

**Expected Behavior:**
The custom integration should handle file operations asynchronously to avoid blocking the event loop, ensuring smooth and responsive operation.

**Actual Behavior:**
The integration attempts to open and read a JSON file synchronously, which blocks the event loop and causes potential delays or disruptions in the Home Assistant's performance.

**Error Log:**
```
Logger: homeassistant.util.loop
Source: util/loop.py:84
First occurred: 14:24:55 (1 occurrences)
Last logged: 14:24:55

Detected blocking call to open inside the event loop by custom integration 'weatherdotcom' at custom_components/weatherdotcom/coordinator.py, line 232: tfiledata = json.load_json(f'{tfiledir}{tfilename}.json') (offender: /usr/src/homeassistant/homeassistant/util/json.py, line 78: with open(filename, mode="rb") as fdesc:), please create a bug report at https://github.com/jaydeethree/Home-Assistant-weatherdotcom/issues/
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 223, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 190, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 672, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 639, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1988, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 165, in async_setup_component
    result = await _async_setup_component(hass, domain, config)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 447, in _async_setup_component
    await asyncio.gather(
  File "/usr/src/homeassistant/homeassistant/setup.py", line 449, in <genexpr>
    create_eager_task(
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 37, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 742, in async_setup_locked
    await self.async_setup(hass, integration=integration)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 594, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/weatherdotcom/__init__.py", line 48, in async_setup_entry
    weathercoordinator = WeatherUpdateCoordinator(hass, config)
  File "/config/custom_components/weatherdotcom/coordinator.py", line 84, in __init__
    self._tranfile = self.get_tran_file()
  File "/config/custom_components/weatherdotcom/coordinator.py", line 232, in get_tran_file
    tfiledata = json.load_json(f'{tfiledir}{tfilename}.json')

```

**Possible Solution:**
Refactor the code to use asynchronous file I/O operations to load the JSON data, thereby preventing the event loop from being blocked.

**Environment:**

Core: 2024.6.2
Supervisor: 2024.06.0
Operating System: 12.3
Frontend: 20240610.0
Integration Version: 1.1.6

**Additional Information:**
This issue impacts the performance of the Home Assistant due to the synchronous nature of the file operation within the event loop. Addressing this will improve the responsiveness and reliability of the integration.

Please let me know if further information is required to resolve this issue.

**Comment by jaydeethree** [comment]
Thanks for the report! This doesn't seem urgent to me (I see similar log messages for several other integrations, including some core integrations), so I'll plan to look at it sometime in the next week or two.

**Comment by jaydeethree** [comment]
I started looking at this to get a better understanding of what I need to change. This error happens because of how I'm loading the JSON file that contains translations, and the right way to fix this is to use Home Assistant's native translation functionality instead of the custom translation handling that's already built into the integration. I started looking at how to switch to that, but it's going to require a lot of changes to this integration and I unfortunately don't have time for that right now. Instead I'm planning to just load the JSON file asynchronously - that will fix this error and be an improvement, but I'd still like to switch to the native translation functionality someday.

**Comment by jaydeethree** [comment]
I spent some time working on this, but unfortunately I haven't been able to fix it yet. I have no experience with Python coroutines/async, and all of the code for that is from https://github.com/cytech/Home-Assistant-wundergroundpws/ which I used as the base for this integration. I tried several different ways of fixing this but all of them created more problems - for example I was able to load the translations asynchronously and make this error go away, but that caused the translations to not be loaded early enough during startup so the sensors didn't have their names set (which led to different, more serious errors).

For now this issue isn't very urgent - it logs an annoying warning, but otherwise it has no real impact. I'll hopefully revisit this at some point in the future, but for now I'm planning to put it on hold. If anyone with more async experience wants to send a PR to fix this I would appreciate it :)

**Comment by Tmatz27** [comment]
Here is my coordinator.py
`
"""The Weather.com data coordinator."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from datetime import datetime, timedelta
import logging
from typing import Any
import json  # Correct import for json

import aiohttp
import async_timeout

from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
from homeassistant.util.unit_system import METRIC_SYSTEM
from homeassistant.const import (
    PERCENTAGE, UnitOfPressure, UnitOfTemperature, UnitOfLength, UnitOfSpeed, UnitOfVolumetricFlux)
from .const import (
    ICON_CONDITION_MAP,
    FIELD_DAYPART,
    FIELD_HUMIDITY,
    FIELD_TEMPERATUREMAX,
    FIELD_TEMPERATUREMIN,
    FIELD_VALIDTIMEUTC,
    FIELD_WINDDIR,
    FIELD_WINDGUST,
    FIELD_WINDSPEED,
    DOMAIN,
    RESULTS_CURRENT,
    RESULTS_FORECAST_DAILY,
    RESULTS_FORECAST_HOURLY
)

_LOGGER = logging.getLogger(__name__)

_RESOURCESHARED = '&format=json&apiKey={apiKey}&units={units}'
_RESOURCECURRENT = ('https://api.weather.com/v3/wx/observations/current'
                    '?geocode={latitude},{longitude}')
_RESOURCEFORECASTDAILY = ('https://api.weather.com/v3/wx/forecast/daily/15day'
                          '?geocode={latitude},{longitude}')
_RESOURCEFORECASTHOURLY = ('https://api.weather.com/v3/wx/forecast/hourly/15day'
                           '?geocode={latitude},{longitude}')

MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=20)


@dataclass
class WeatherUpdateCoordinatorConfig:
    """Class representing coordinator configuration."""

    api_key: str
    location_name: str
    unit_system_api: str
    unit_system: str
    lang: str
    latitude: str
    longitude: str
    update_interval = MIN_TIME_BETWEEN_UPDATES


class WeatherUpdateCoordinator(DataUpdateCoordinator):
    """The Weather.com update coordinator."""

    icon_condition_map = ICON_CONDITION_MAP

    def __init__(
            self, hass: HomeAssistant, config: WeatherUpdateCoordinatorConfig
    ) -> None:
        """Initialize."""
        self._hass = hass
        self._api_key = config.api_key
        self._location_name = config.location_name
        self._unit_system_api = config.unit_system_api
        self.unit_system = config.unit_system
        self._lang = config.lang
        self._latitude = config.latitude
        self._longitude = config.longitude
        self.data = None
        self._session = async_get_clientsession(self._hass)
        self._tranfile = None
        self._tranfile_loaded_event = asyncio.Event()  # Add event for translation file loading

        if self._unit_system_api == 'm':
            self.units_of_measurement = (UnitOfTemperature.CELSIUS, UnitOfLength.MILLIMETERS, UnitOfLength.METERS,
                                         UnitOfSpeed.KILOMETERS_PER_HOUR, UnitOfPressure.MBAR,
                                         UnitOfVolumetricFlux.MILLIMETERS_PER_HOUR, PERCENTAGE)
            self.visibility_unit = UnitOfLength.KILOMETERS
        else:
            self.units_of_measurement = (UnitOfTemperature.FAHRENHEIT, UnitOfLength.INCHES, UnitOfLength.FEET,
                                         UnitOfSpeed.MILES_PER_HOUR, UnitOfPressure.INHG,
                                         UnitOfVolumetricFlux.INCHES_PER_HOUR, PERCENTAGE)
            self.visibility_unit = UnitOfLength.MILES

        super().__init__(
            hass,
            _LOGGER,
            name="WeatherUpdateCoordinator",
            update_interval=config.update_interval,
        )
        self._hass.loop.create_task(self.async_get_tran_file())

    @property
    def is_metric(self):
        """Determine if this is the metric unit system."""
        return self._hass.config.units is METRIC_SYSTEM

    @property
    def location_name(self):
        """Return the location used for data."""
        return self._location_name

    async def _async_update_data(self) -> dict[str, Any]:
        return await self.get_weather()

    async def get_weather(self):
        """Get weather data."""
        headers = {
            'Accept-Encoding': 'gzip',
            "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36"
        }
        try:
            with async_timeout.timeout(10):
                url = self._build_url(_RESOURCECURRENT)
                response = await self._session.get(url, headers=headers)
                result_current = await response.json(content_type=None)
                if result_current is None:
                    raise ValueError('NO CURRENT RESULT')
                self._check_errors(url, result_current)

            with async_timeout.timeout(10):
                url = self._build_url(_RESOURCEFORECASTDAILY)
                response = await self._session.get(url, headers=headers)
                result_forecast_daily = await response.json(content_type=None)

                if result_forecast_daily is None:
                    raise ValueError('NO DAILY FORECAST RESULT')
                self._check_errors(url, result_forecast_daily)

            with async_timeout.timeout(10):
                url = self._build_url(_RESOURCEFORECASTHOURLY)
                response = await self._session.get(url, headers=headers)
                result_forecast_hourly = await response.json(content_type=None)

                if result_forecast_hourly is None:
                    raise ValueError('NO HOURLY FORECAST RESULT')
                self._check_errors(url, result_forecast_hourly)

            result = {
                RESULTS_CURRENT: result_current,
                RESULTS_FORECAST_DAILY: result_forecast_daily,
                RESULTS_FORECAST_HOURLY: result_forecast_hourly,
            }

            self.data = result

            return result

        except ValueError as err:
            _LOGGER.error("Check Weather.com API %s", err.args)
            raise UpdateFailed(err)
        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
            _LOGGER.error("Error fetching Weather.com data: %s", repr(err))
            raise UpdateFailed(err)
        # _LOGGER.debug(f'Weather.com data {self.data}')

    def _build_url(self, baseurl):
        baseurl += '&language={language}'
        baseurl += _RESOURCESHARED

        return baseurl.format(
            apiKey=self._api_key,
            language=self._lang,
            latitude=self._latitude,
            longitude=self._longitude,
            units=self._unit_system_api
        )

    def _check_errors(self, url: str, response: dict):
        # _LOGGER.debug(f'Checking errors from {url} in {response}')
        if 'errors' not in response:
            return
        if errors := response['errors']:
            raise ValueError(
                f'Error from {url}: '
                '; '.join([
                    e['message']
                    for e in errors
                ])
            )

    def get_current(self, field):
        return self.data[RESULTS_CURRENT][field]

    def get_forecast_daily(self, field, period=0):
        try:
            if field in [
                FIELD_TEMPERATUREMAX,
                FIELD_TEMPERATUREMIN,
                FIELD_VALIDTIMEUTC,
            ]:
                # Those fields exist per-day, rather than per dayPart, so the period is halved
                return self.data[RESULTS_FORECAST_DAILY][field][int(period / 2)]
            return self.data[RESULTS_FORECAST_DAILY][FIELD_DAYPART][0][field][period]
        except IndexError:
            return None

    def get_forecast_hourly(self, field, hour):
        try:
            return self.data[RESULTS_FORECAST_HOURLY][field][hour]
        except IndexError:
            return None

    @classmethod
    def _iconcode_to_condition(cls, icon_code):
        for condition, iconcodes in cls.icon_condition_map.items():
            if icon_code in iconcodes:
                return condition
        _LOGGER.warning(f'Unmapped iconCode from TWC Api. (44 is Not Available (N/A)) "{icon_code}". ')
        return None

    @classmethod
    def _format_timestamp(cls, timestamp_secs):
        return datetime.utcfromtimestamp(timestamp_secs).isoformat('T') + 'Z'

    async def async_get_tran_file(self):
        """Get translation file for Weather.com sensor friendly_name"""
        tfiledir = f'{self._hass.config.config_dir}/custom_components/{DOMAIN}/weather_translations/'
        tfilename = self._lang.split('-', 1)[0]
        loop = asyncio.get_running_loop()
        try:
            tfiledata = await loop.run_in_executor(None, self._load_json, f'{tfiledir}{tfilename}.json')
        except Exception:  # pylint: disable=broad-except
            tfiledata = await loop.run_in_executor(None, self._load_json, f'{tfiledir}en.json')
            _LOGGER.warning(f'Sensor translation file {tfilename}.json does not exist. Defaulting to en-US.')
        self._tranfile = tfiledata
        self._tranfile_loaded_event.set()  # Signal that translation file is loaded

    async def wait_for_tranfile(self):
        """Wait until the translation file is loaded."""
        await self._tranfile_loaded_event.wait()

    @staticmethod
    def _load_json(filepath):
        """Load JSON file"""
        with open(filepath, 'r') as f:
            return json.load(f)  # Correct usage of json.load


class InvalidApiKey(HomeAssistantError):
    """Error to indicate there is an invalid api key."""
`

here is my updated sensor.py
`"""
Sensor Support for Weather.com weather service.
For more details about this platform, please refer to the documentation at
https://github.com/jaydeethree/Home-Assistant-weatherdotcom
"""
from __future__ import annotations

import logging

from homeassistant.components.sensor import SensorEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity import generate_entity_id
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity
from homeassistant.util.unit_system import METRIC_SYSTEM

from .coordinator import WeatherUpdateCoordinator

from .const import (
    CONF_ATTRIBUTION,
    DOMAIN,
    FIELD_DAYPART,
    FIELD_WINDGUST,
    FIELD_WINDSPEED,
    RESULTS_CURRENT,
    RESULTS_FORECAST_DAILY,
    RESULTS_FORECAST_HOURLY
)
from .weather_current_conditions_sensors import *

_LOGGER = logging.getLogger(__name__)

# Declaration of supported Weather.com observation/condition sensors
SENSOR_DESCRIPTIONS: tuple[WeatherSensorEntityDescription, ...] = (
    current_condition_sensor_descriptions
)


async def async_setup_entry(
        hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Add Weather.com entities from a config_entry."""
    coordinator: WeatherUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
    await coordinator.wait_for_tranfile()  # Ensure translation file is loaded
    sensors = [
        WeatherSensor(coordinator, description) for description in SENSOR_DESCRIPTIONS
    ]

    async_add_entities(sensors)


class WeatherSensor(CoordinatorEntity, SensorEntity):
    """Implementing the Weather.com sensor."""
    _attr_has_entity_name = True
    _attr_attribution = CONF_ATTRIBUTION
    entity_description: WeatherSensorEntityDescription

    def __init__(
            self,
            coordinator: WeatherUpdateCoordinator,
            description: WeatherSensorEntityDescription,
    ):
        super().__init__(coordinator)
        self.entity_description = description

        entity_id_format = description.key + ".{}"

        self._attr_unique_id = f"{self.coordinator.location_name},{description.key}".lower()
        self.entity_id = generate_entity_id(
            entity_id_format, f"{self.coordinator.location_name}_{description.name}", hass=coordinator.hass
        )
        self._unit_system = coordinator.unit_system
        if description.key == 'latitude':
            self._sensor_data = coordinator._latitude
        elif description.key == 'longitude':
            self._sensor_data = coordinator._longitude
        else:
            self._sensor_data = _get_sensor_data(
                coordinator.data, description.key, self._unit_system)
        self._attr_native_unit_of_measurement = self.entity_description.unit_fn(
            self.coordinator.hass.config.units is METRIC_SYSTEM)

    @property
    def available(self) -> bool:
        """Return if weather data is available."""
        return self.coordinator.data is not None

    @property
    def name(self):
        """Return the name of the sensor."""
        # Ensure the translation file is loaded before accessing it
        if self.coordinator._tranfile is not None:
            if self.entity_description.key in self.coordinator._tranfile.keys() or \
                    self.entity_description.key in self.coordinator._tranfile[FIELD_DAYPART].keys():
                return self.coordinator._tranfile[self.entity_description.key]

        return self.entity_description.name

    @property
    def native_value(self) -> StateType:
        """Return the state."""
        return self.entity_description.value_fn(self._sensor_data, self._unit_system)

    @property
    def extra_state_attributes(self) -> dict[str, Any]:
        """Return the state attributes."""
        return self.entity_description.attr_fn(self.coordinator.data)

    @callback
    def _handle_coordinator_update(self) -> None:
        """Handle data update."""
        if self.entity_description.key == 'latitude':
            self._sensor_data = self.coordinator._latitude
        elif self.entity_description.key == 'longitude':
            self._sensor_data = self.coordinator._longitude
        else:
            self._sensor_data = _get_sensor_data(
                self.coordinator.data, self.entity_description.key, self._unit_system
            )
        self.async_write_ha_state()


def _get_sensor_data(
        sensors: dict[str, Any],
        kind: str,
        unit_system: str
) -> Any:
    """Get sensor data."""
    # windGust is often null. When it is, set it to windSpeed instead.
    if kind == FIELD_WINDGUST and sensors[RESULTS_CURRENT][kind] == None:
        return sensors[RESULTS_CURRENT][FIELD_WINDSPEED]
    else:
        return sensors[RESULTS_CURRENT][kind]
`
This seems to have fix the issue for me

**Comment by Tmatz27** [comment]
not sure why it put the code in there like that ill try and put it in a file to upload

**Comment by jaydeethree** [comment]
Thanks so much! I should have time to test that out next week, and if everything looks good I'll push a new release with the changes.

**Comment by Tmatz27** [comment]
Not a problem I seriously appreciated the work you did and just wanted to help where I could.

[sensor.txt](https://github.com/user-attachments/files/16200438/sensor.txt)
[coordinator.txt](https://github.com/user-attachments/files/16200439/coordinator.txt)
 Attached as a txt file since it got messed up above and can't post .pay files in here

**Comment by jaydeethree** [comment]
I started testing your fix and it looks great, thank you so much! I also noticed that the other integration that this one is based on added a fix for this error recently:

https://github.com/cytech/Home-Assistant-wundergroundpws/commit/36e33bf17e07fa342e1e66d9ab007382d5b1b9ea

I'm going to take a closer look at their fix and then decide whether to use theirs or yours, but I'm leaning towards using their fix simply because that's what my integration is forked from. Keeping the integrations similar to each other will make it easier for us to share fixes in the future when other issues come up.

I **really** appreciate your help with this even if I don't end up using your fix - thanks so much!
